### PR TITLE
Packer 0.7.5

### LINKED ISSUE #72: errors installing driver_restapi
go get github.com/MSOpenTech/packer-azure/packer/builder/azure/driver_restapi

# github.com/MSOpenTech/packer-azure/packer/builder/azure/driver_restapi
../../MSOpenTech/packer-azure/packer/builder/azure/driver_restapi/builder.go:44: undefined: packer.ConfigTemplate

I have tried with binary packer 0.7.2 and 0.7.5, compiled dev version (0.7.5) and still get the same error.
Also tried golang 1.3 and 1.4

Using CentOS 7, will all updates as of 8/6

really need to get this working since we working to get our product into Azure via packer.

### PR DESCRIPTION
**Pull request by paulmey**
fixes #72, doing some testing before merging

cc: @ahmetalpbalkan @sriramsa @gabrielhartmann

### CODE REVIEW

File: packer/builder/azure/driver_restapi/config.go [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+
+	if c.UserImageLabel == "" {
+		errs = packer.MultiErrorAppend(errs, fmt.Errorf("user_image_label must be specified"))
+	}
+
+	// check
+	pattern := "^[A-Za-z][A-Za-z0-9-_.]*[A-Za-z0-9]$"
+	value := c.UserImageLabel
+
+	match, _ := regexp.MatchString(pattern, value)
```
Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/config.go]: don't ignore err.

File: packer/builder/azure/driver_restapi/config.go [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+	if !sizeIsValid {
+		errs = packer.MultiErrorAppend(errs, fmt.Errorf("instance_size is not valid, must be one of: %v", instanceSizeArr))
+	}
+
+	if c.UserImageLabel == "" {
+		errs = packer.MultiErrorAppend(errs, fmt.Errorf("user_image_label must be specified"))
+	}
+
+	// check
+	pattern := "^[A-Za-z][A-Za-z0-9-_.]*[A-Za-z0-9]$"
```
Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/config.go]: p :=  regexp.MustCompile(`"^[A-Za-z][A-Za-z0-9-_.]*[A-Za-z0-9]$"`)
    ...
   p.MatchString

notice the backticks as well

File: packer/builder/azure/driver_restapi/config.go [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+	value := c.UserImageLabel
```
Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/config.go]: no need for new var

File: packer/builder/azure/driver_restapi/config.go [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+)
+
+type Config struct {
+	common.PackerConfig `mapstructure:",squash"`
+
+	SubscriptionName        string `mapstructure:"subscription_name"`
+	PublishSettingsPath     string `mapstructure:"publish_settings_path"`
+	StorageAccount          string `mapstructure:"storage_account"`
+	StorageAccountContainer string `mapstructure:"storage_account_container"`
+	OsType                  string `mapstructure:"os_type"`
```
Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/config.go]: `OS` per go conventions, fwiw

File: packer/builder/azure/driver_restapi/config.go [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+		errs = packer.MultiErrorAppend(errs, fmt.Errorf("publish_settings_path is not a valid path: %s", err))
+	}
+
+	osTypeIsValid := false
+	osTypeArr := []string{
+		targets.Linux,
+		targets.Windows,
+	}
+
+	for _, osType := range osTypeArr {
```
Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/config.go]: loop seems redundant. why not 

    if !(osType == targets.Linux || osType == targets.Windows) already

File: packer/builder/azure/driver_restapi/config.go [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+	if c.OsImageLabel == "" {
+		errs = packer.MultiErrorAppend(errs, fmt.Errorf("os_image_label must be specified"))
+	}
+
+	if c.Location == "" {
+		errs = packer.MultiErrorAppend(errs, fmt.Errorf("location must be specified"))
+	}
+
+	sizeIsValid := false
+	instanceSizeArr := []string{
```
Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/config.go]: you can make this method more readable my exporting this into a pkg level var.

File: packer/builder/azure/driver_restapi/config.go [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+		targets.Standard_DS13,
+		targets.Standard_DS14,
+		targets.Standard_G1,
+		targets.Standard_G2,
+		targets.Standard_G3,
+		targets.Standard_G4,
+		targets.Standard_G5,
+	}
+
+	for _, instanceSize := range instanceSizeArr {
```
Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/config.go]: in case when you find yourself in a position where you check multiple times if a `string` is in `[]string`,  you can write a method 

    func contains(s []string, v string) bool

File: packer/builder/azure/driver_restapi/config.go [hunk 8]
```diff
@@ -0,0 +1,200 @@
...
+
+	c.userImageName = utils.DecorateImageName(c.UserImageLabel)
+	c.tmpContainerName = utils.BuildContainerName()
+
+	// for Win  - the computer name cannot be more than 15 characters long
+	const tmpServiceNamePrefix = "PkrSrv"
+	const tmpVmNamePrefix = "PkrVM"
+	randSuffix := utils.BuildAzureVmNameRandomSuffix(tmpVmNamePrefix)
+
+	if c.tmpVmName == "" {
```
Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/config.go]: these 3 `if` statements actually do one thing, if you like, you can have them all single line and you'll end up 3 short lines instead of 12

**Comment by ahmetalpbalkan** [comment]
generally lgtm, have comments on some nits.

File: packer/builder/azure/driver_restapi/config.go [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+	PublishSettingsPath     string `mapstructure:"publish_settings_path"`
+	StorageAccount          string `mapstructure:"storage_account"`
+	StorageAccountContainer string `mapstructure:"storage_account_container"`
+	OsType                  string `mapstructure:"os_type"`
+	OsImageLabel            string `mapstructure:"os_image_label"`
+	Location                string `mapstructure:"location"`
+	InstanceSize            string `mapstructure:"instance_size"`
+	UserImageLabel          string `mapstructure:"user_image_label"`
+
+	username         string `mapstructure:"username"`
```
Comment (sriramsa) [inline packer/builder/azure/driver_restapi/config.go]: username should be 'userName' ?

File: packer/builder/azure/driver_restapi/config.go (hunk 8 above)
Comment (sriramsa) [inline packer/builder/azure/driver_restapi/config.go]: Since c isn't returned if err has a value, should we check err and not bother updating tmp* values in c in the next few lines?. Might be alright if this is needed for debugging/logging.

File: packer/builder/azure/driver_restapi/config.go (hunk 5 above)
Comment (paulmey) [inline packer/builder/azure/driver_restapi/config.go]: cut-paste :+1:

File: packer/builder/azure/driver_restapi/config.go [hunk 10]
```diff
@@ -0,0 +1,200 @@
...
+
+	if c.tmpServiceName == "" {
+		c.tmpServiceName = tmpServiceNamePrefix + randSuffix
+	}
+
+	if c.username == "" {
+		c.username = "packer"
+	}
+
+	log.Println(fmt.Sprintf("%s: %v", "subscription_name", c.SubscriptionName))
```
Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/config.go]: hmm this is too long. Why not just serialize into JSON or something? or `fmt.Sprintf("%#v")`?

File: packer/builder/azure/driver_restapi/config.go (hunk 7 above)
Comment (paulmey) [inline packer/builder/azure/driver_restapi/config.go]: well, you reduced this to the only instance... I think :wink:

**Comment by paulmey** [comment]
pushed some deltas, will squash before merge

File: packer/builder/azure/driver_restapi/config.go (hunk 10 above)
Comment (paulmey) [inline packer/builder/azure/driver_restapi/config.go]: that is an option... but I want to do that in a separate PR...

File: packer/builder/azure/driver_restapi/config.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+
+type Config struct {
+	common.PackerConfig `mapstructure:",squash"`
+
+	SubscriptionName        string `mapstructure:"subscription_name"`
+	PublishSettingsPath     string `mapstructure:"publish_settings_path"`
+	StorageAccount          string `mapstructure:"storage_account"`
+	StorageAccountContainer string `mapstructure:"storage_account_container"`
+	OSType                  string `mapstructure:"os_type"`
+	OsImageLabel            string `mapstructure:"os_image_label"`
```
Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/config.go]: Os → OS

File: packer/builder/azure/driver_restapi/config.go [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+	"regexp"
+)
+
+type Config struct {
+	common.PackerConfig `mapstructure:",squash"`
+
+	SubscriptionName        string `mapstructure:"subscription_name"`
+	PublishSettingsPath     string `mapstructure:"publish_settings_path"`
+	StorageAccount          string `mapstructure:"storage_account"`
+	StorageAccountContainer string `mapstructure:"storage_account_container"`
```
Comment (ahmetalpbalkan) [inline packer/builder/azure/driver_restapi/config.go]: Probably should be called StorageContainer, but ignore if it was an existing setting.

Comment (paulmey) [inline packer/builder/azure/driver_restapi/config.go]: yeah, I can fix the property on the struct but not the json name...

**Comment by paulmey** [comment]
some more fixes

**Comment by paulmey** [comment]
squashed, ready to merge
